File: Serenity.Script.UI/FilterPanel/FilterWidgetBase.cs
using jQueryApi;
using System;
using System.Collections.Generic;
using System.Runtime.CompilerServices;

namespace Serenity
{
    [IncludeGenericArguments(false), ScriptName("FilterWidgetBase")]
    public abstract class FilterWidgetBase<TOptions> : TemplatedWidget<TOptions>
        where TOptions: class, new()
    {
        private FilterStore store;

        public FilterWidgetBase(jQueryObject div, TOptions opt)
            : base(div, opt)
        {
            this.store = new FilterStore(new List<PropertyItem>());
            this.store.Changed += OnFilterStoreChanged;
        }

        public override void Destroy()
        {
            if (store != null)
            {
                store.Changed -= OnFilterStoreChanged;
                store = null;
            }

            base.Destroy();
        }

        private void OnFilterStoreChanged(object sender, EventArgs e)
        {
            FilterStoreChanged();
        }

        protected virtual void FilterStoreChanged()
        {
        }

        public FilterStore Store
        {
            get { return store; }
            set
            {
                if (store != value)
                {
                    if (this.store != null)
                        this.store.Changed -= OnFilterStoreChanged;

                    store = value ?? new FilterStore(new List<PropertyItem>());
                    store.Changed += OnFilterStoreChanged;
                    FilterStoreChanged();
                }
            }
        }
    }
}

File: src/Serenity.Scripts/CoreLib/UI/Filtering/FilterWidgetBase.ts
import { Decorators } from "../../Decorators";
import { TemplatedWidget } from "../Widgets/TemplatedWidget";
import { FilterStore } from "./FilterStore";

@Decorators.registerClass('Serenity.FilterWidgetBase')
export class FilterWidgetBase<TOptions> extends TemplatedWidget<TOptions> {
    private store: FilterStore;

    private onFilterStoreChanged: () => void;

    constructor(div: JQuery, opt?: TOptions) {
        super(div, opt);

        this.store = new FilterStore([]);
        this.onFilterStoreChanged = () => this.filterStoreChanged();
        this.store.add_changed(this.onFilterStoreChanged);
    }

    destroy() {

        if (this.store) {
            this.store.remove_changed(this.onFilterStoreChanged);
            this.onFilterStoreChanged = null;
            this.store = null;
        }

        super.destroy();
    }

    protected filterStoreChanged() {
    }

    get_store(): FilterStore {
        return this.store;
    }

    set_store(value: FilterStore): void {
        if (this.store !== value) {
            if (this.store != null) 
                this.store.remove_changed(this.onFilterStoreChanged);

            this.store = value || new FilterStore([]);
            this.store.add_changed(this.onFilterStoreChanged);
            this.filterStoreChanged();
        }
    }
}